File: RoomInn/app/src/main/java/postpc/finalproject/RoomInn/Launch/LaunchActivity.java
package postpc.finalproject.RoomInn.Launch;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentContainerView;
import androidx.lifecycle.ViewModelProvider;

import com.facebook.CallbackManager;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

import postpc.finalproject.RoomInn.MainActivity;
import postpc.finalproject.RoomInn.R;
import postpc.finalproject.RoomInn.ViewModle.LoginViewModel;

public class LaunchActivity extends AppCompatActivity {
    int RC_SIGN_IN = 0;

    private Fragment loginFragment;
    private Fragment registerFragment;
    private LoginViewModel viewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.getWindow().getDecorView().setLayoutDirection(View.LAYOUT_DIRECTION_LTR);
        setContentView(R.layout.activity_launch);
        FragmentContainerView fragment = findViewById(R.id.fragment_frame);

        viewModel = new ViewModelProvider(this).get(LoginViewModel.class);

        loginFragment = new LoginFragment();
        registerFragment = new RegisterFragment();

        getSupportFragmentManager().beginTransaction()
                .replace(fragment.getId(), loginFragment)
                .commit();
    }

    // TODO: uncomment this! (do not delete this piece of code!)
    @Override
    protected void onStart() {
        // this function will activate if the user signed up with google before
        super.onStart();
        GoogleSignInAccount acct = GoogleSignIn.getLastSignedInAccount(this);
        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);
//        getToMainActivity();
    }

    private void getToMainActivity() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach a listener.
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            handleSignInResult(task);
        }

        // Result returned from launching the Intent from FacebookActivity;
        else {
            viewModel.getCallbackManager().onActivityResult(requestCode, resultCode, data);
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    private void handleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            GoogleSignInAccount account = completedTask.getResult(ApiException.class);

            // Signed in successfully, show authenticated UI.
            getToMainActivity();
        } catch (ApiException e) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.w("Error", "signInResult:failed code=" + e.getStatusCode());
        }
    }
}

File: RoomInn/app/src/main/java/postpc/finalproject/RoomInn/launch/LaunchActivity.kt
package postpc.finalproject.RoomInn.launch

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentContainerView
import androidx.lifecycle.ViewModelProvider
import com.facebook.AccessToken
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import postpc.finalproject.RoomInn.MainActivity
import postpc.finalproject.RoomInn.R
import postpc.finalproject.RoomInn.ViewModle.LoginViewModel
import postpc.finalproject.RoomInn.ViewModle.ProjectViewModel
import postpc.finalproject.RoomInn.models.RoomInnApplication.Companion.getInstance

class LaunchActivity : AppCompatActivity() {
    var RC_SIGN_IN = 0
    private var progressDialog: ProgressDialog? = null
    private var mGoogleSignInClient: GoogleSignInClient? = null
    lateinit private var titleText: TextView

    private val viewModel: LoginViewModel by lazy {
        ViewModelProvider(this).get(LoginViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.window.decorView.layoutDirection = View.LAYOUT_DIRECTION_LTR
        setContentView(R.layout.activity_launch)
        titleText = findViewById(R.id.MyProjects_title)

        viewModel.googleSingInCallable =  { googleSignIn() }
        viewModel.changeTitle = { titleText.text = it }

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build()
        // Build a GoogleSignInClient with the options specified by gso.
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso)



        val fragment = findViewById<FragmentContainerView>(R.id.fragment_frame)
        if (viewModel.isRegistering) {
            supportFragmentManager.beginTransaction()
                    .replace(fragment.id, viewModel.registerFragment)
                    .commit()
        } else {
            supportFragmentManager.beginTransaction()
                    .replace(fragment.id, viewModel.loginFragment)
                    .commit()
        }
    }

    override fun onStart() {
        super.onStart()
        val userId = userID()
        if (userId != null) {
            progressDialog = ProgressDialog(this)
            progressDialog!!.setTitle("Login")
            progressDialog!!.setMessage("Please wait")
            progressDialog!!.setCanceledOnTouchOutside(false)
            progressDialog!!.show()
            FirebaseFirestore.getInstance().collection("users").document(userId).get()
                    .addOnSuccessListener { v: DocumentSnapshot ->
                        if (v.exists()) {
                            getToMainActivity(userId)
                            progressDialog!!.dismiss()
                            Toast.makeText(this, "Login successful", Toast.LENGTH_SHORT)
                                    .show()
                        } else {
                            FirebaseAuth.getInstance().signOut()
                            GoogleSignIn.getClient(applicationContext, GoogleSignInOptions.DEFAULT_SIGN_IN).signOut()
                            progressDialog!!.dismiss()
                            Toast.makeText(this, "Login failed", Toast.LENGTH_SHORT)
                                    .show()
                        }
                    }
                    .addOnFailureListener { v: Exception? ->
                        FirebaseAuth.getInstance().signOut()
                        GoogleSignIn.getClient(applicationContext, GoogleSignInOptions.DEFAULT_SIGN_IN).signOut()
                        progressDialog!!.dismiss()
                        Toast.makeText(this, "Login failed", Toast.LENGTH_SHORT)
                                .show()
                    }
        } else {
            FirebaseAuth.getInstance().signOut()
        }
    }

    private fun getToMainActivity(userId: String) {
        getInstance().getRoomsDB().initialize(userId)
        val intent = Intent(this, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
        startActivity(intent)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RC_SIGN_IN) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            handleSignInResult(task)
        } else {
            viewModel.callbackManager.onActivityResult(requestCode, resultCode, data)
        }
    }

    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
        try {
            val account = completedTask.getResult(ApiException::class.java)

            // Signed in successfully, show authenticated UI.
            getToMainActivity(account.id!!)
        } catch (e: ApiException) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.w("Error", "signInResult:failed code=" + e.statusCode)
        }
    }

    // user  is already logged in with facebook
    private fun userID(): String? {
            var userId: String? = null
            val acct = GoogleSignIn.getLastSignedInAccount(this)
            if (acct != null) {
                userId = acct.id
                Log.d("login", "login with google id: $userId")
            }

            // user  is already logged in with facebook
            val accessToken = AccessToken.getCurrentAccessToken()
            val isLoggedIn = accessToken != null && !accessToken.isExpired
            if (isLoggedIn) {
                userId = accessToken!!.userId
                Log.d("login", "login with facebook id: $userId")
            }
            val user = FirebaseAuth.getInstance().currentUser
            if (user != null) {
                userId = user.uid
                Log.d("login", "login with firebase id: $userId")
            }
            return userId
        }

    override fun onBackPressed() {
        if (viewModel.isRegistering) {
            viewModel.isRegistering = false
            val ft = supportFragmentManager.beginTransaction()
                    .replace(findViewById<View>(R.id.fragment_frame).id, viewModel.loginFragment)
            ft.commit()
        } else {
            super.onBackPressed()
        }
    }

    private fun googleSignIn() {
        val signInIntent = mGoogleSignInClient!!.signInIntent
        signInIntent.putExtra("rc_code", RC_SIGN_IN)
        startActivityForResult(signInIntent, RC_SIGN_IN)
    }
}